aloha-core/src/main/scala/com/eharmony/aloha/audit/impl/EitherAuditor.scala

```diff
@@ -0,0 +1,175 @@
+      modelId: ModelIdentity,
+      errorMsgs: => Seq[String] = Nil,
+      missingVarNames: => Set[String] = Set.empty,
+      subvalues: Seq[Either[EitherAuditorError, _]] = Nil
+  ): Either[EitherAuditorError, A] = {
+    val (errors, missing, failIds) = aggregateDiagnostics(errorMsgs, missingVarNames, subvalues)
+    Left(EitherAuditorError(::(modelId, failIds), errors, missing))
+  }
+
+  override def success(
+      modelId: ModelIdentity,
+      valueToAudit: A,
+      errorMsgs: => Seq[String] = Nil,
+      missingVarNames: => Set[String] = Set.empty,
+      subvalues: Seq[Either[EitherAuditorError, _]] = Nil,
+      prob: => Option[Float] = None
+  ): Either[EitherAuditorError, A] = Right(valueToAudit)
+
+  /**
+    * Construct a new EitherAuditor[M].  Notice this allows succeeds.
+    * @tparam M the type of success value in the new auditor.
+    * @return
+    */
+  protected[this] def changeTo[M]: EitherAuditor[M]
+
+  /**
+    * (''Possibly'') aggregate the information from `subvalues` into the error messages, missing
+    * variable names and model IDs with problems
+    * @param errorMsgs error messages from the current model failure.
+    * @param missingVarNames missing variable names from the current model failure.
+    * @param subvalues diagnostic information from submodel successes / failures.
+    * @return (''possibly'') aggregated errors, missing variable info, and IDs of submodels that
+    *         couldn't produce predictions.
+    */
+  protected[this] def aggregateDiagnostics(
+      errorMsgs: Seq[String],
+      missingVarNames: Set[String],
+      subvalues: Seq[Either[EitherAuditorError, _]]
+  ): (Seq[String], Set[String], List[ModelIdentity])
+}
+
+object EitherAuditor {
+
+  /**
+    * Produce an [[EitherAuditor]] that aggregates diagnostic information.
+    * @tparam A the type of value returned in the success case.
+    * @return
+    */
+  def apply[A]: EitherAuditor[A] = apply[A](aggregateDiagnostics = true)
+
+  /**
+    * Produce an [[EitherAuditor]].
+    * @param aggregateDiagnostics  If `true`, aggregate diagnostics; otherwise,
+    *                              disregard submodel information in the
+    *                              [[EitherAuditorError]] returned by the auditor.
+    * @tparam A the type of value returned in the success case.
+    * @return
+    */
+  def apply[A](aggregateDiagnostics: Boolean): EitherAuditor[A] =
+    if (aggregateDiagnostics)
+      AggEitherAuditor[A]()
+    else NoAggEitherAuditor[A]()
+
+  private[this] final case class AggEitherAuditor[A]() extends EitherAuditor[A] {
+    override protected[this] def changeTo[M]: EitherAuditor[M] = AggEitherAuditor[M]()
+    override protected[this] def aggregateDiagnostics(
+        err: Seq[String],
+        missing: Set[String],
+        sub: Seq[Either[EitherAuditorError, _]]): (Seq[String], Set[String], List[ModelIdentity]) = {
+      val e = sub.foldLeft(err)((s, sv) => s ++ sv.fold(e => e.errorMsgs, _ => Nil))
```

Comment:
These can't be `flatMap`.  Must be `foldLeft`.  Otherwise, there are type issues.